clang-tools-extra/clangd/refactor/Rename.cpp

```diff
@@ -538,11 +565,254 @@ std::optional<InvalidName> checkName(const NamedDecl &RenameDecl,
+            halfOpenToRange(SM, Tok.range(SM).toCharRange(SM)));
+        continue;
+      }
+    }
+
+    switch (Tok.kind()) {
+    case tok::l_square:
+      if (parseMessageExpression(Tokens, SM, Index + 1, Last, *Selector,
+                                 SelectorPieces)) {
+        Ranges.emplace_back(std::move(SelectorPieces));
+        SelectorPieces.clear();
+      }
+      break;
+    case tok::l_paren:
+      ParenCount++;
+      break;
+    case tok::r_paren:
+      if (ParenCount > 0) {
+        --ParenCount;
+      }
+      break;
+    case tok::r_brace:
+      if (BraceCount > 0) {
+        --BraceCount;
+      }
+      break;
+    case tok::l_brace:
+      // At the top level scope we should only have method defs.
+      if (BraceCount == 0 && ParenCount == 0) {
+        // All the selector pieces in the method def have been found.
+        if (SelectorPieces.size() == NumArgs) {
+          Ranges.emplace_back(std::move(SelectorPieces));
+        }
+        SelectorPieces.clear();
+      }
+      ++BraceCount;
+      break;
+    case tok::semi:
+      // At the top level scope we should only have method decls.
+      if (BraceCount == 0 && ParenCount == 0) {
+        // All the selector pieces in the method decl have been found.
+        if (SelectorPieces.size() == NumArgs) {
+          Ranges.emplace_back(std::move(SelectorPieces));
+        }
+        SelectorPieces.clear();
+      }
+      break;
+    default:
+      break;
+    }
+  }
+  return Ranges;
+}
+
+clangd::Range tokenRangeForLoc(ParsedAST &AST, SourceLocation TokLoc,
+                               const SourceManager &SM,
+                               const LangOptions &LangOpts) {
+  const auto *Token = AST.getTokens().spelledTokenAt(TokLoc);
+  assert(Token && "rename expects spelled tokens");
+  clangd::Range Result;
+  Result.start = sourceLocToPosition(SM, Token->location());
+  Result.end = sourceLocToPosition(SM, Token->endLocation());
   return Result;
 }
 
+// AST-based ObjC method rename, it renames all occurrences in the main file
+// even for selectors which may have multiple tokens.
+llvm::Expected<tooling::Replacements>
+renameObjCMethodWithinFile(ParsedAST &AST, const ObjCMethodDecl *MD,
+                           llvm::StringRef NewName,
+                           std::vector<SourceLocation> SelectorOccurences) {
+  const SourceManager &SM = AST.getSourceManager();
+  auto Code = SM.getBufferData(SM.getMainFileID());
+  auto RenameIdentifier = MD->getSelector().getNameForSlot(0).str();
+  llvm::SmallVector<llvm::StringRef, 8> NewNames;
+  NewName.split(NewNames, ":");
+
+  std::vector<Range> Ranges;
+  const auto &LangOpts = MD->getASTContext().getLangOpts();
+  for (const auto &Loc : SelectorOccurences)
+    Ranges.push_back(tokenRangeForLoc(AST, Loc, SM, LangOpts));
+  auto FilePath = AST.tuPath();
+  auto RenameRanges = collectRenameIdentifierRanges(
+      RenameIdentifier, Code, LangOpts, MD->getSelector());
```

Comment:
Am I missing something here or is this doing a purely syntactic rename, ignoring `SelectorOccurrences` altogether?

Reply:
Yeah, initially I did have a custom AST visitor to do a proper semantic rename but we decided to stick with the same approach as the global rename.

Reply:
That seems quite error-prone to me. If we have index data that tells us where the selectors appear, we should be using it in my opinion – both for local and global rename.